services/src/datasets/external/netcdfcf/gdalmd.rs

```diff
@@ -0,0 +1,547 @@
+            let buffer_stride: *const i64 = std::ptr::null(); // default value
+            let data_type = GDALMDArrayGetDataType(c_mdarray);
+            let p_dst_buffer_alloc_start: *mut c_void = std::ptr::null_mut();
+            let n_dst_buffer_alloc_size = 0;
+
+            let rv = gdal_sys::GDALMDArrayRead(
+                c_mdarray,
+                array_start_index.as_ptr(),
+                count.as_ptr(),
+                array_step,
+                buffer_stride,
+                data_type,
+                string_pointers.as_mut_ptr().cast::<std::ffi::c_void>(),
+                p_dst_buffer_alloc_start,
+                n_dst_buffer_alloc_size,
+            );
+
+            if rv == 0 {
+                return Err(GdalError::BadArgument("GDALMDArrayRead failed".to_string()));
+            }
+
+            let strings = string_pointers
+                .into_iter()
+                .map(|string_ptr| {
+                    let string = Self::_string(string_ptr);
+
+                    VSIFree(string_ptr as *mut c_void);
+
+                    string
+                })
+                .collect();
+
+            GDALExtendedDataTypeRelease(data_type);
+
+            GDALMDArrayRelease(c_mdarray);
+
+            GDALReleaseDimensions(c_dimensions, dim_count);
+
+            strings
+        };
+
+        Ok(value)
+    }
+
+    pub fn group_names(&self) -> Vec<String> {
+        let options = CslStringList::new();
+
+        unsafe {
+            let c_group_names = GDALGroupGetGroupNames(self.c_group, options.as_ptr());
+
+            let strings = Self::_string_array(c_group_names);
+
+            CSLDestroy(c_group_names);
+
+            strings
+        }
+    }
+
+    pub fn array_names(&self) -> Vec<String> {
+        let mut options = CslStringList::new();
+        options
+            .set_name_value("SHOW_ZERO_DIM", "NO")
+            .unwrap_or_else(|e| debug!("{}", e));
+        options
+            .set_name_value("SHOW_COORDINATES", "NO")
+            .unwrap_or_else(|e| debug!("{}", e));
+        options
+            .set_name_value("SHOW_INDEXING", "NO")
+            .unwrap_or_else(|e| debug!("{}", e));
+        options
+            .set_name_value("SHOW_BOUNDS", "NO")
+            .unwrap_or_else(|e| debug!("{}", e));
+        options
+            .set_name_value("SHOW_TIME", "NO")
+            .unwrap_or_else(|e| debug!("{}", e));
+        options
+            .set_name_value("GROUP_BY", "SAME_DIMENSION")
+            .unwrap_or_else(|e| debug!("{}", e));
+
+        unsafe {
+            let c_group_names = GDALGroupGetMDArrayNames(self.c_group, options.as_ptr());
+
+            let strings = Self::_string_array(c_group_names);
+
+            CSLDestroy(c_group_names);
+
+            strings
+        }
+    }
+
+    pub fn open_group(&self, name: &str) -> Result<Self> {
+        let c_name = CString::new(name).expect("no null-byte in name");
```

Comment:
Unsafeness and being able to panic are two different things.

In my GDAL PR I handle this differently and this will replace this codeblock. Moreover, this was not changed by this PR. I just moved the methods.